### PR TITLE
feat: implement validateInitialSchemaNode with state-based validation

### PR DESCRIPTION
**Pull request by FunamaYukina**
## Summary

This PR implements the `validateInitialSchemaNode` functionality with significant improvements to the architecture:

- ✨ Added `isEmptySchema` utility to `@liam-hq/schema` package for reusability
- 🔄 Refactored `validateInitialSchemaNode` to use `state.schemaData` instead of external DB calls
- 🚦 Implemented conditional workflow routing based on first execution detection
- 🧪 Added comprehensive integration tests for validation scenarios

## Architecture
```mermaid
%%{init: {'flowchart': {'curve': 'linear'}}}%%
graph TD;
	__start__([<p>__start__</p>]):::first
	validateInitialSchema(validateInitialSchema)
	leadAgent(leadAgent)
	pmAgent(pmAgent)
	dbAgent(dbAgent)
	qaAgent(qaAgent)
	__end__([<p>__end__</p>]):::last
	dbAgent --> qaAgent;
	pmAgent --> dbAgent;
	qaAgent --> leadAgent;
	validateInitialSchema --> leadAgent;
	__start__ -.-> validateInitialSchema;
	__start__ -.-> leadAgent;
	leadAgent -.-> pmAgent;
	leadAgent -.-> __end__;
	classDef default fill:#f2f0ff,line-height:1.2;
	classDef first fill-opacity:0;
	classDef last fill:#bfb6fc;
```

### 1. Schema Utility Enhancement
- Created `isEmptySchema` utility function in `@liam-hq/schema` package
- Added comprehensive unit tests covering all edge cases
- Exported from package for monorepo-wide reusability

### 2. State-Based Validation 
- Removed external DB dependency (`repositories.schema.getSchema`)
- Now uses `state.schemaData` directly for validation
- Improved testability and reduced I/O operations
- Follows LangGraph best practices for state management

### 3. Workflow Routing
- Added conditional edge from START node
- Routes to `validateInitialSchema` on first execution (no AI messages)
- Routes directly to `leadAgent` for subsequent executions
- Ensures validation only runs once per conversation

### 4. Error Handling
- Implemented `createValidationError` helper for consistent error creation
- Used `ResultAsync` pattern from neverthrow for proper error handling
- Simplified error message extraction using `JSON.stringify`

## Testing

The UI (error message display) hasn't been designed yet, but it will look like this:

**Success case**
https://liam-app-git-feat-validate-initial-schema-logic-re-liambx.vercel.app/app/design_sessions/53901995-4051-41c3-8a89-4a639ee30c49?deepModeling=true

ss 3849

 **failure case**
https://liam-app-git-feat-validate-initial-schema-logic-re-liambx.vercel.app/app/design_sessions/7a6aa005-d515-4231-9fcf-210e8800feae?deepModeling=true

ss 3848

## Next PR: Error Message UI Integration

The next PR will focus on:
- **Error message persistence**: Save validation errors to timeline_item for UI display
- **User interface improvements**: Design and implement user-friendly error message display in the chat interface
- **Error message streaming**: Ensure validation errors are properly surfaced to users through the existing message streaming system

## Migration Notes

The node now uses `state.schemaData` instead of fetching from the database. This is a non-breaking change as the state is already populated by previous nodes in the workflow.

🤖 Generated with [Claude Code](https://claude.ai/code)

## Summary by CodeRabbit

* **New Features**
  * Smarter workflow routing: first run performs initial database schema validation; subsequent runs skip it for faster responses.
  * Robust schema validation with clear, user-facing error reporting when schema deparsing or execution fails.
  * New utility to detect empty schemas, improving behavior on fresh setups.

* **Tests**
  * Added integration-style workflow tests and unit tests for schema-empty detection, validation scenarios, and error handling.

* **Documentation**
  * Updated architecture diagram to show conditional start routing.

### CODE REVIEW

File: frontend/internal-packages/agent/src/workflow/nodes/validateInitialSchemaNode.integration.test.ts [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+        latestVersionNumber: 1,
+        next: 'leadAgent',
+      }
+
+      // Act
+      const stream = await graph.stream(state, config)
+
+      // Assert (Output)
+      await outputStream(stream)
+    })
```
[the remark this answers is not in the document]
Comment (FunamaYukina) [reply frontend/internal-packages/agent/src/workflow/nodes/validateInitialSchemaNode.integration.test.ts]: @coderabbitai
This is a test used for debugging purposes.So it's fine as it is.

File: frontend/internal-packages/agent/src/workflow/nodes/validateInitialSchemaNode.integration.test.ts [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
```
[the remark this answers is not in the document]
Comment (FunamaYukina) [reply frontend/internal-packages/agent/src/workflow/nodes/validateInitialSchemaNode.integration.test.ts]: @coderabbitai
This is a test used for debugging purposes.So it's fine as it is.

File: frontend/internal-packages/agent/src/workflow/nodes/validateInitialSchemaNode.integration.test.ts [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+  })
```
[the remark this answers is not in the document]
Comment (FunamaYukina) [reply frontend/internal-packages/agent/src/workflow/nodes/validateInitialSchemaNode.integration.test.ts]: @coderabbitai
Thank you. This integration test was intended for debugging purposes, but since it's difficult to write tests that verify errors, I've decided to remove it.
https://github.com/liam-hq/liam/pull/3359/commits/08de051bf9e3dec3ccda24cacae67dcd54d30a57

File: frontend/internal-packages/agent/src/workflow/nodes/validateInitialSchemaNode.ts [hunk 4]
```diff
@@ -1,14 +1,60 @@
...
+import { ResultAsync } from 'neverthrow'
 import type { WorkflowState } from '../../chat/workflow/types'
+import { WorkflowTerminationError } from '../../shared/errorHandling'
+
+const createValidationError = (message: string): WorkflowTerminationError => {
+  return new WorkflowTerminationError(
+    new Error(message),
+    'validateInitialSchemaNode',
+  )
+}
```
[the remark this answers is not in the document]
Comment (FunamaYukina) [reply frontend/internal-packages/agent/src/workflow/nodes/validateInitialSchemaNode.ts]: @coderabbitai
Thank you! I fixed it.
https://github.com/liam-hq/liam/pull/3359/commits/1c2c3ebcf2f11dfa0bcc1fdf14b0cde497cdabab

File: frontend/internal-packages/agent/src/workflow/nodes/validateInitialSchemaNode.ts [hunk 5]
```diff
@@ -2,13 +1,60 @@
...
+    (error) => new Error(String(error)),
+  ).match(
+    (validationResults) => {
+      const hasErrors = validationResults.some((result) => !result.success)
+
+      if (hasErrors) {
+        const errorResult = validationResults.find((result) => !result.success)
+        const errorMessage = JSON.stringify(errorResult?.result)
+        throw createValidationError(`Schema validation failed: ${errorMessage}`)
+      }
```
[the remark this answers is not in the document]
Comment (FunamaYukina) [reply frontend/internal-packages/agent/src/workflow/nodes/validateInitialSchemaNode.ts]: @coderabbitai
I believe it has improved with the commit below.
https://github.com/liam-hq/liam/commit/1c2c3ebcf2f11dfa0bcc1fdf14b0cde497cdabab

File: frontend/internal-packages/agent/src/workflow/nodes/validateInitialSchemaNode.ts [hunk 6]
```diff
@@ -2,13 +1,60 @@
...
+
+      // TODO: Add message creation in next PR
+      return state
+    },
+    (error) => {
+      throw createValidationError(
+        `Schema validation execution failed: ${error.message}`,
+      )
```
[the remark this answers is not in the document]
Comment (FunamaYukina) [reply frontend/internal-packages/agent/src/workflow/nodes/validateInitialSchemaNode.ts]: @coderabbitai
I believe it has improved with the commit below.
https://github.com/liam-hq/liam/pull/3359/commits/13c77feb7d575b32d907c60d31617050a2d50568

File: frontend/internal-packages/agent/src/workflow/nodes/validateInitialSchemaNode.ts [hunk 7]
```diff
@@ -1,14 +1,61 @@
...
 export async function validateInitialSchemaNode(
   state: WorkflowState,
-  _config: RunnableConfig,
 ): Promise<WorkflowState> {
-  // TODO: Implement validation logic
-  return state
+  if (isEmptySchema(state.schemaData)) {
+    // TODO: Add message creation in next PR
+    return state
+  }
```
Comment (MH4GF) [inline frontend/internal-packages/agent/src/workflow/nodes/validateInitialSchemaNode.ts:26]: I thought this node would be simpler if this check was implemented with conditonal edges.

File: frontend/internal-packages/agent/src/workflow/nodes/validateInitialSchemaNode.ts [hunk 8]
```diff
@@ -1,14 +1,61 @@
...
+}
 
 /**
  * Validates initial schema and provides Instant Database initialization experience.
  * Only runs on first workflow execution.
  */
 export async function validateInitialSchemaNode(
   state: WorkflowState,
-  _config: RunnableConfig,
 ): Promise<WorkflowState> {
```
[the remark this answers is not in the document]
Comment (FunamaYukina) [reply frontend/internal-packages/agent/src/workflow/nodes/validateInitialSchemaNode.ts:22]: The config: RunnableConfig parameter is needed only when the node requires:
  - Repository access (const { repositories } = configurableResult.value)
  - External service calls (LLM agents, APIs)
  - Context information (thread_id, organizationId, etc.)

  validateInitialSchemaNode doesn't need config because it:
  - Only performs pure schema validation
  - No external resource access required
  - Works solely with state.schemaData

File: frontend/internal-packages/agent/src/workflow/nodes/validateInitialSchemaNode.ts [hunk 9]
```diff
@@ -1,14 +1,61 @@
...
+
+  const ddlStatements = ddlResult.value
+  const requiredExtensions = Object.keys(
+    state.schemaData.extensions || {},
+  ).sort()
+
+  return await ResultAsync.fromPromise(
+    executeQuery(ddlStatements, requiredExtensions),
+    (error) => (error instanceof Error ? error : new Error(String(error))),
+  ).match(
```
Comment (MH4GF) [inline frontend/internal-packages/agent/src/workflow/nodes/validateInitialSchemaNode.ts]: Since executeQuery does try-catch internally and does not throw errors, I think it is ok to skip error handling with `ResultAsync.fromAsyncPromise`. 
https://github.com/supermacro/neverthrow?tab=readme-ov-file#resultasyncfromsafepromise-static-class-method

File: frontend/internal-packages/agent/src/workflow/nodes/validateInitialSchemaNode.integration.test.ts
Comment (MH4GF) [inline frontend/internal-packages/agent/src/workflow/nodes/validateInitialSchemaNode.integration.test.ts:1]: The integration test is operated for the purpose of easily executing the unstable LLM API. This time the node does not execute LLM and will always return the same result. How about doing `expect()` as a regular `ValidateInitialSchemaNode.test.ts`?

File: frontend/internal-packages/agent/src/workflow/nodes/validateInitialSchemaNode.ts [hunk 10]
```diff
@@ -1,14 +1,61 @@
...
+    (validationResults) => {
+      const hasErrors = validationResults.some((result) => !result.success)
+
+      if (hasErrors) {
+        const errorResult = validationResults.find((result) => !result.success)
+        const errorMessage = JSON.stringify(errorResult?.result)
```
[the remark this answers is not in the document]
Comment (FunamaYukina) [reply frontend/internal-packages/agent/src/workflow/nodes/validateInitialSchemaNode.ts]: I'd like to see what actual error structures we encounter in practice before implementing a custom formatting solution. We can revisit this once we have more real-world error data to inform the best approach.

File: frontend/internal-packages/agent/src/workflow/nodes/validateInitialSchemaNode.ts (hunk 7 above)
Comment (FunamaYukina) [reply frontend/internal-packages/agent/src/workflow/nodes/validateInitialSchemaNode.ts:26]: I agree, but I prioritized communicating that “the instance DB environment is ready” even if the schema is empty. (I'll be developing the UI in the next pull request, so I think we can decide there.)

Comment (MH4GF) [reply frontend/internal-packages/agent/src/workflow/nodes/validateInitialSchemaNode.ts:26]: Oh, that's a good idea. I think it is good to put it here because it displays a message!

File: frontend/internal-packages/agent/src/workflow/nodes/validateInitialSchemaNode.ts (hunk 9 above)
Comment (FunamaYukina) [reply frontend/internal-packages/agent/src/workflow/nodes/validateInitialSchemaNode.ts]: Oh, you're absolutely right! Thank you!!
Removed `ResultAsync.fromPromise` since executeQuery returns results with success/failure indicators rather than throwing exceptions.

https://github.com/liam-hq/liam/pull/3359/commits/644f51b4c2412221746f300568aab93edc652325

File: frontend/internal-packages/agent/src/workflow/nodes/validateInitialSchemaNode.integration.test.ts
Comment (FunamaYukina) [reply frontend/internal-packages/agent/src/workflow/nodes/validateInitialSchemaNode.integration.test.ts:1]: You were right about that too. By writing it using expect() like a regular test, I was able to add tests for error cases as well!
https://github.com/liam-hq/liam/pull/3359/commits/6177d418225843061c1b2126f8932c0d4f85ffa3
https://github.com/liam-hq/liam/pull/3359/commits/e5a002e0da604ba2cae69dc3ea80a8d3c5b215ac

File: frontend/internal-packages/agent/src/workflow/nodes/validateInitialSchemaNode.ts (hunk 9 above)
Comment (MH4GF) [reply frontend/internal-packages/agent/src/workflow/nodes/validateInitialSchemaNode.ts]: Nice! We don't need to use neverthrow here. The code is much easier to read now!

File: frontend/internal-packages/agent/src/workflow/nodes/validateInitialSchemaNode.integration.test.ts [hunk 11]
```diff
@@ -0,0 +1,270 @@
...
+        latestVersionNumber: 1,
+        next: 'leadAgent',
+      }
+
+      // Act
+      const stream = await graph.stream(state, config)
+      const results = []
+      for await (const result of stream) {
+        results.push(result)
+      }
```
Comment (MH4GF) [inline frontend/internal-packages/agent/src/workflow/nodes/validateInitialSchemaNode.integration.test.ts]: In this case, since we don't need to output to a stream for debugging, I think it's clearer to receive the state via invoke. The following code is conceptual and has not been tested.

```suggestion
      const result = await graph.invoke(state, config)
      expect(result).toEqual(state) // Nothing has changed at present. Messages will be added in the future.
```

File: frontend/internal-packages/agent/src/workflow/nodes/validateInitialSchemaNode.integration.test.ts
Comment (MH4GF) [reply frontend/internal-packages/agent/src/workflow/nodes/validateInitialSchemaNode.integration.test.ts:1]: Thank you! To align with the naming convention for other tests, please rename it to `validateInitialSchemaNode.test.ts`.

File: frontend/internal-packages/agent/src/workflow/nodes/validateInitialSchemaNode.integration.test.ts (hunk 11 above)
Comment (FunamaYukina) [reply frontend/internal-packages/agent/src/workflow/nodes/validateInitialSchemaNode.integration.test.ts]: You were absolutely right!
 I've made the corrections below.
https://github.com/liam-hq/liam/pull/3359/commits/6c7fe800a1b382e328dae2fa8c8d2396d2adf046

File: frontend/internal-packages/agent/src/workflow/nodes/validateInitialSchemaNode.integration.test.ts
Comment (FunamaYukina) [reply frontend/internal-packages/agent/src/workflow/nodes/validateInitialSchemaNode.integration.test.ts:1]: Thank you!  I have changed the file name to ``validateInitialSchemaNode.test.ts``.
https://github.com/liam-hq/liam/pull/3359/commits/6c7fe800a1b382e328dae2fa8c8d2396d2adf046

**Comment by FunamaYukina** [comment]
Thank you for the review!!

File: frontend/internal-packages/agent/src/workflow/nodes/validateInitialSchemaNode.test.ts [hunk 12]
```diff
@@ -1,9 +1,262 @@
...
+  return {
+    ...actual,
+    isEmptySchema: (schema: Parameters<typeof actual.isEmptySchema>[0]) => {
+      return (
+        Object.keys(schema.tables || {}).length === 0 &&
+        Object.keys(schema.enums || {}).length === 0
+      )
+    },
+  }
+})
```
Comment (MH4GF) [inline frontend/internal-packages/agent/src/workflow/nodes/validateInitialSchemaNode.test.ts:22]: Why was mocking necessary?

File: frontend/internal-packages/agent/src/workflow/nodes/validateInitialSchemaNode.test.ts [hunk 13]
```diff
@@ -1,9 +1,262 @@
...
-    expect(typeof validateInitialSchemaNode).toBe('function')
+// Mock OPENAI_API_KEY for getTestConfig
+// validateInitialSchemaNode doesn't use OpenAI API, so dummy key is sufficient
+beforeAll(() => {
+  process.env['OPENAI_API_KEY'] = 'dummy-key-for-testing'
+})
+
+afterAll(() => {
+  delete process.env['OPENAI_API_KEY']
+})
```
Comment (MH4GF) [inline frontend/internal-packages/agent/src/workflow/nodes/validateInitialSchemaNode.test.ts:34]: I see. This seems fine for now, but if the number increases, it might be better to modify the helper instead.

File: frontend/internal-packages/agent/src/workflow/nodes/validateInitialSchemaNode.test.ts [hunk 14]
```diff
@@ -1,9 +1,262 @@
...
+      // Assert
+      // Schema validation should succeed with valid schema
+      expect(result).toEqual(state)
+
+      // Should have users table with correct structure
+      expect(result.schemaData.tables['users']).toBeDefined()
+      const usersTable = result.schemaData.tables['users']
+      expect(usersTable?.name).toBe('users')
+      expect(usersTable?.columns['id']?.type).toBe('uuid')
+      expect(usersTable?.columns['email']?.type).toBe('varchar')
```
Comment (MH4GF) [inline frontend/internal-packages/agent/src/workflow/nodes/validateInitialSchemaNode.test.ts:138]: Why is this assertion necessary? Isn't it already covered by `expect(result).toEqual(state)`?

File: frontend/internal-packages/agent/src/workflow/nodes/validateInitialSchemaNode.test.ts [hunk 15]
```diff
@@ -1,9 +1,262 @@
...
+      const result = await graph.invoke(state, config)
+
+      // Assert
+      // Non-first execution should return state unchanged (validation is skipped)
+      expect(result).toEqual(state)
+      expect(result.userInput).toBe('Follow-up message')
+
+      // Should preserve the existing messages including AI message
+      expect(result.messages).toHaveLength(3)
+      expect(result.messages[1]).toBeInstanceOf(AIMessage)
```
Comment (MH4GF) [inline frontend/internal-packages/agent/src/workflow/nodes/validateInitialSchemaNode.test.ts:200]: Why is this assertion necessary? Isn't it already covered by `expect(result).toEqual(state)`?

File: frontend/internal-packages/agent/src/workflow/nodes/validateInitialSchemaNode.test.ts [hunk 16]
```diff
@@ -1,9 +1,262 @@
...
+      )
+      expect(thrownError).toBeInstanceOf(WorkflowTerminationError)
+
+      const errorMessage =
+        thrownError instanceof WorkflowTerminationError
+          ? thrownError.message
+          : String(thrownError)
+      expect(errorMessage).toBe(
+        'Error in validateInitialSchemaNode: Schema validation failed: {"error":"type \\"unknown_invalid_type\\" does not exist"}',
+      )
```
Comment (MH4GF) [inline frontend/internal-packages/agent/src/workflow/nodes/validateInitialSchemaNode.test.ts:259]: This can be verified in one go.

```suggestion
      await expect(validateInitialSchemaNode(state)).rejects.toThrow(
        new WorkflowTerminationError('Error in validateInitialSchemaNode: Schema validation failed: {"error":"type \\"unknown_invalid_type\\" does not exist"}'),
      )
```

https://vitest.dev/api/expect.html#tothrowerror

**Review by MH4GF — COMMENTED** [review]
@FunamaYukina 
Sorry, there are a few areas in the test that concern me, so please address them if possible.

**Comment by FunamaYukina** [comment]
@MH4GF Thank you for the review🙏
I will address this in a separate pull request.
